### ISSUE TITLE
Feature Request: Enhanced Auto-Parallel, Auto-Placement, and Auto-Topology Support for Verl (Improvement of Hybridflow Paper)

### ISSUE THREAD

**Issue by aminmalekisadr** [volcengine/verl#4328]
### Feature request

 Feature Request: Auto-Parallel, Auto-Placement, and Auto-Topology Search for RLHF Training

1.	Summary

Request to add comprehensive auto-optimization support to VERL for RLHF training, including:
1. Auto-Parallel: Automatically determining optimal parallelism degrees (Data Parallel, Tensor Parallel, Pipeline Parallel, Expert Parallel, Sequence Parallel, Context Parallel)
2. Auto-Placement: Automatically finding optimal device placement for models (actor, critic, reference, reward) across available GPUs
3. Auto-Topology Search: Automatically optimizing topology-aware placement considering rack topology, inter-node vs intra-node bandwidth, and communication patterns

This feature supports PPO, GRPO, DAPO, and DPO algorithms.

### Motivation

2.	Motivation

Currently, users must manually configure three critical aspects of RLHF training:
1. Parallelism Degrees: `tensor_model_parallel_size`, `pipeline_model_parallel_size`, `data_parallel_size`, etc.
2. Device Placement: Which models (actor, critic, reference, reward) go on which GPUs, and also which type of collocation (split, nemo,collocate, etc)? 
3. Topology-Aware Placement: How to place models considering rack topology and network bandwidth

This requires:
- Deep understanding of model architecture, hardware topology, and communication patterns
- Time-consuming manual tuning for each model/hardware combination
- Knowledge of inter-node vs intra-node bandwidth implications

Comprehensive auto-optimization would:
- Automatically find optimal parallelism degrees based on model size, GPU count, and topology
- Automatically determine best device placement for all models
- Automatically optimize topology-aware placement considering rack structure and bandwidth
- Reduce configuration complexity for users
- Improve resource utilization and training efficiency
- Support multiple RLHF algorithms (PPO, GRPO, DAPO, DPO)

### Your contribution

Proposed Solution

3.	Components

1. Auto-Parallel Search Engine
   - Exhaustive search over valid parallelism degree combinations (DP, TP, PP, EP, SP, CP)
   - Memory and compute time estimation for each configuration
   - Support for FSDP/FSDP2 and DeepSpeed ZeRO sharding strategies
   - Micro-batch size optimization

2. Auto-Placement Search Engine
   - search over all possible device placement configurations
   - Evaluates co-location vs separation strategies for models
-	Optimum number of GPU/NPUs for each model
   - Considers memory constraints and communication overhead
   - Optimizes for end-to-end training time (generation + inference + training + resharding)

3. Auto-Topology Search Engine
   - Topology-aware placement optimization considering rack structure
   - Inter-node vs intra-node bandwidth analysis
   - Communication pattern optimization (inter, intra node for different parallelization)
   - Integration with Ray placement groups for launcher
   - Symmetric vs sequential placement strategies

4. Integration Points
   - integration: Add `--auto-parallel` flag to training scripts
   - Config integration: Add `auto_parallel` section to config files
   - Search-only mode: Run search and output optimal config (no training)
   - Launcher integration: Apply topology-aware placement during job launch

5. Algorithm Support
   - PPO (Proximal Policy Optimization) - actor, critic, reference, reward
   - GRPO (Group Relative Policy Optimization) - actor, reference, reward
   - DAPO (Decoupled Clip and Dynamic sAmpling Policy Optimization) - actor, reference, reward
   - DPO (Direct Preference Optimization) - actor, reference only

4.	Architecture

```
User Config
    ↓
Comprehensive Auto-Optimization Search
    ├── Auto-Parallel Search
    │   ├── Model Analysis (memory, compute)
    │   ├── Degree Generation (DP, TP, PP, EP, SP, CP)
    │   └── Micro-batch Size Optimization
    ├── Auto-Placement Search
    │   ├── Placement Generation (co-locate vs separate)
    │   ├── Memory Constraint Checking
    │   └── Communication Overhead Analysis
    └── Auto-Topology Search
        ├── Topology Analysis (rack structure, bandwidth)
        ├── Inter-node vs Intra-node Optimization
        └── Ray Placement Group Generation
    ↓
Output: Optimal Configuration
    ├── Parallelism Degrees
    │   ├── tensor_model_parallel_size
    │   ├── pipeline_model_parallel_size
    │   ├── data_parallel_size
    │   ├── expert_model_parallel_size
    │   ├── sequence_parallel (bool)
    │   ├── context_parallel_size
    │   └── micro_batch_size
    ├── Device Placement
    │   ├── Actor placement
    │   ├── Critic placement (PPO only)
    │   ├── Reference placement
    │   └── Reward placement
    └── Topology-Aware Placement
        ├── Selected nodes/racks
        ├── Placement group configuration
        └── Bandwidth allocation strategy
```

5.	Implementation Details

1. Configuration Interface

**Config File Format:**
```yaml
rl_config:
  auto_parallel:
    enabled: true  # Enable comprehensive auto-optimization
    world_size: 16  # Total number of GPUs
    gpu_name: "v100-sxm-32gb"  # GPU type for analysis
    topo_path: "avail_topo"  # Topology file path (for auto-topology search)
    sharding_strategy: "fsdp"  # "none", "fsdp", "fsdp2", "ds_zero_1/2/3"
    opt_parallel: true  # Enable auto-parallel: optimize parallelism degrees
    opt_placement: true  # Enable auto-placement: optimize device placement
    optimize_topology: true  # Enable auto-topology: optimize topology-aware placement
    rlhf_algo: "grpo"  # Algorithm: "ppo", "grpo", "dapo", "dpo"
    factor: 2  # GPU allocation granularity
```

```

6.	Search Results Format:
```python
{
    'optimal_generate_config': {
        'pipeline_model_parallel_size': 2,
        'tensor_model_parallel_size': 4,
        'data_parallel_size': 2,
        'rl_config/actor_rollout_dispatch_size': 1,
    },
    'optimal_ref_config': {
        'pipeline_model_parallel_size': 2,
        'tensor_model_parallel_size': 4,
        'context_parallel_size': 1,
        'rl_config/ref_forward_micro_batch_size': 1,
    },
    'optimal_actor_config': {
        'pipeline_model_parallel_size': 2,
        'tensor_model_parallel_size': 4,
        'megatron_training/micro_batch_size': 1,
    }
}
```

8.	Benefits

1. Ease of Use: Users don't need to manually tune parallelism degrees, device placement, or topology configuration
2. Performance: Automatically finds near-optimal configurations for all three dimensions
3. Flexibility: Supports multiple RLHF algorithms and sharding strategies
4. Topology-Aware: Considers hardware topology, rack structure, and bandwidth for optimal placement
5. Memory Efficient: Accounts for FSDP/ZeRO sharding strategies
6. Communication Optimized: Minimizes inter-node communication and maximizes intra-node bandwidth utilization
7. End-to-End Optimization: Optimizes entire training pipeline (generation + inference + training + resharding)

9.	Files Structure
```
verl/trainer/auto_parallel/
├── __init__.py
├── rlhf_auto_placement.py  # Main search interface (orchestrates all three)
├── rlhf_auto_parallel.py   # Auto-parallel: core parallelism degree search
├── grpo_simu.py            # GRPO simulation (placement + parallel)
├── ppo_simu.py             # PPO simulation (placement + parallel)
├── dapo_simu.py            # DAPO simulation (placement + parallel)
├── dpo_simu.py             # DPO simulation (placement + parallel)
├── model.py                # Model analysis wrapper
├── rlhf_topo_search.py     # Auto-topology: topology-aware placement
├── utils.py                # Utility functions
└── llm_analysis/           # Memory/compute analysis
   

```

 After (Comprehensive Auto-Optimization)
```yaml
rl_config:
  auto_parallel:
    enabled: true
    world_size: 16
    gpu_name: "v100-sxm-32gb"
    topo_path: "avail_topo"
    rlhf_algo: "grpo"
    sharding_strategy: "fsdp"
    opt_parallel: true      # Auto-parallel
    opt_placement: true    # Auto-placement
    optimize_topology: true # Auto-topology
```

10.	Testing Plan

1. Unit tests for degree generation and validation
2. Integration tests for search algorithm
3. End-to-end tests with different model sizes and GPU counts
4. Performance benchmarks comparing auto vs manual configurations

11.	Future Enhancements

1. Adaptive Auto-Optimization: Adjust degrees, placement, and topology during training based on runtime metrics. 
2. Dynamic Topology: Adapt to changing cluster topology during training
3. Multi Agent RLHF considerations

Priority: High  
Effort: (3-4 weeks for comprehensive implementation)  
Impact: Very High (significantly improves user experience, resource utilization, and training performance)

[Feature Request Verl.docx](https://github.com/user-attachments/files/23810471/Feature.Request.Verl.docx)

**Comment by Kirrito-k423** [comment]
This issue is being closed by the scheduled stale-issue cleanup because it was created at 2025-11-27T20:15:53Z, has 0 comments, and the checked timeline did not show any public references, cross-references, mentions, connected events, or other public linked interactions. The issue has therefore had no public interaction for more than 6 months before the 2026-07-24 run. Evidence checked: issue page https://github.com/verl-project/verl/issues/4328, comments API https://api.github.com/repos/verl-project/verl/issues/4328/comments (count: 0), and timeline API https://api.github.com/repos/verl-project/verl/issues/4328/timeline (no public reference events found). If this is still relevant, please comment with a current reproduction, evidence, or requirement context so maintainers can reopen it, or open a new issue with updated details.

本 issue 按定时清理规则关闭：它创建于 2025-11-27T20:15:53Z，评论数量为 0，并且已检查的 timeline 中未发现公开引用、PR/issue cross-reference、mention、connected 或其它公开关联互动；因此截至 2026-07-24 本轮运行已超过 6 个月没有公开互动。已检查证据：issue 页面 https://github.com/verl-project/verl/issues/4328、comments API https://api.github.com/repos/verl-project/verl/issues/4328/comments（数量 0）、timeline API https://api.github.com/repos/verl-project/verl/issues/4328/timeline（未发现公开引用事件）。如果该问题仍然有效，请带上当前版本的复现、证据或需求背景重新评论，维护者可以据此重开；也可以创建包含最新信息的新 issue。

This response was generated by a scheduled Codex GPT5.5 issue processing task.
